projects/01-rubygems-cli/solution/lib/api/ruby_gems_api_with_cache.rb

```diff
@@ -0,0 +1,42 @@
+require "./lib/cache/db_cache"
+require "./lib/api/ruby_gems_api"
+
+class RubyGemsAPIWithCache
+  class << self
+    SHOW_ACTION = "show"
+    SEARCH_ACTION = "search"
+
+    def show_gem(gem_name)
+      key = cache_key(gem_name, SHOW_ACTION)
+      entry = cache.read(key)
+      return GemData.new(JSON.parse(entry.value)) if entry
+
+      gem = RubyGemsAPI.show_gem(gem_name)
+      cache.write(key, JSON.dump(gem.raw_data))
+
+      gem
+    end
+
+    def search_gems(keyword)
+      key = cache_key(keyword, SEARCH_ACTION)
+      entry = cache.read(key)
+      return JSON.parse(entry.value).map { |gem| GemData.new(gem) } if entry
+
+      gems = RubyGemsAPI.search_gems(keyword)
+      raw_data = gems.map { |gem| gem.raw_data }
+      cache.write(key, JSON.dump(raw_data))
+
+      gems
```

Comment:
Very good. Here's a challenge: try rewriting this using [tap](https://ruby-doc.org/core-2.6.1/Object.html#method-i-tap). It's a useful method that allows you to do some neat stuff once you get the hang of it.